src/redis/windowed_cache_impl.go

```diff
@@ -0,0 +1,263 @@
+package redis
+
+import (
+	"math"
+	"math/rand"
+
+	"github.com/coocood/freecache"
+	pb "github.com/envoyproxy/go-control-plane/envoy/service/ratelimit/v3"
+	"github.com/envoyproxy/ratelimit/src/assert"
+	"github.com/envoyproxy/ratelimit/src/config"
+	"github.com/envoyproxy/ratelimit/src/limiter"
+	"github.com/envoyproxy/ratelimit/src/utils"
+	"github.com/golang/protobuf/ptypes/duration"
+	logger "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+type windowedRateLimitCacheImpl struct {
+	client Client
+	// Optional Client for a dedicated cache of per second limits.
+	// If this client is nil, then the Cache will use the client for all
+	// limits regardless of unit. If this client is not nil, then it
+	// is used for limits that have a SECOND unit.
+	perSecondClient            Client
+	timeSource                 limiter.TimeSource
+	jitterRand                 *rand.Rand
+	expirationJitterMaxSeconds int64
+	cacheKeyGenerator          limiter.CacheKeyGenerator
+	localCache                 *freecache.Cache
+	nearLimitRatio             float32
+}
+
+func maxInt64(a int64, b int64) int64 {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func minInt64(a int64, b int64) int64 {
+	if a < b {
+		return a
+	}
+	return b
+}
```

Comment:
The fact that this is implemented here and not in the std lib makes me cry, but so it goes. If this is not present elsewhere in the code can you at least put it in a common utility? If it is already elsewhere please share it?

Reply:
Moving max and min related code to utils